feat: Document how to integrate Mender with Okta (#1976)
mendersoftware/mender-docs: 08.Server-integration/08.SAML-Federated-Authentication/01.OKTA/docs.md

---
title: Okta
taxonomy:
    category: docs
---

!!!!! SAML Federated Authentication is only available in the Mender Enterprise plan.
!!!!! See [the Mender features page](https://mender.io/product/features?target=_blank)
!!!!! for an overview of all Mender plans and features.

> This example will show you how to set Okta as an IdP (Identity Provider) for Hosted Mender.

Reviewer: ```suggestion
This example will show you how to set Okta as an Identity Provider (IdP) for Hosted Mender.
```

The example is based on the free trial accounts on both services.

Create a free trial account for [Okta](https://www.okta.com/free-trial/) and [Hosted Mender](https://mender.io/demo).

## Initial setup on the Okta side

In the Okta Admin panel go to 

> `Directory` -> `People` -> `Add Person` ([Okta - add users](https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-add-users.htm)).

Reviewer: I think this needs a little textual explanation too. Just the screenshot makes context hard to skim read and impossible to index. Plus, if one of the fields has a special meaning it needs to be noted.

Author: You're right.
This is a deliberate decision.
I don't want to be better than documenting OKTA than [OKTA themselves](https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-add-users.htm) as it's expensive to get these images right (even for these simple ones I had to go back and forth multiple times). 

Not sure what you mean by `impossible to index`.

Reviewer: Impossible to index means that search engines can't look into it. So the decision is effectively hindering discoverability of the documentation. Please revisit.

Author: I appreciate your comments, but this is slowly turning into bit rot and I'm not finding the time to improve it.

Going with what @MuchoLucho and @lluiscampos suggested - we'll be merging as in a 'better then nothing' form

![Okta create user](okta_create_user.png)

In the Okta Admin panel go to `Applications` -> `Applications` -> `Create App Integration`.
Select `SAML 2.0` as the Sign-in method and proceed.

> In the `General Settings` name the app `Hosted Mender` and proceed.

Reviewer 2: ```suggestion
In the `General Settings`, set the app's name to `Hosted Mender` and proceed.
```

> In the `General Settings`, set the app's name to `Hosted Mender` and proceed.

Reviewer: Is "Hosted Mender" mandatory, or can the name be chosen by the user?

![Okta create app](okta_create_app_1.png)

> Set the following in the  `SAML Settings`, section `General`:

Reviewer 2: ```suggestion
Set the following values in the `SAML Settings`, section `General`:
```

|         Field                                    |   Value          |
|--------------------------------------------------|------------------|
| `Single sign-on URL`                             | `http://tmp.url` |
| `Use this for Recipient URL and Destination URL` | `CHECKED`        |
| `Audience URI (SP Entity ID)`                    | `http://tmp.url` |
| `Name ID format`                                 | `EmailAddress`   |
| `Application username`                           | `Email`          |

> The `http://tmp.url` are temporary values we'll later replace with correct ones. 

Reviewer 2: ```suggestion
The `http://tmp.url` are temporary values we'll later replace with the correct ones. 
```

This is a workaround to get Okta to generate the IdP metadata XML file.

> Once the content of that XML file is added to Hosted Mender, it will generate the correct URLs.

Reviewer 2: ```suggestion
Once you add the content of that XML to Hosted Mender, it will generate the correct URLs you will replace in Okta.
```

Proceed to the next screen.

> [Assign mender to user](https://help.okta.com/en-us/Content/Topics/Provisioning/lcm/lcm-assign-app-user.htm)

Reviewer: Again, needs text.

![Okta create app](okta_create_app_2.png)

> Set the following in the  `SAML Settings`, section `General`:

Reviewer 2: ```suggestion
Set the following values in the `SAML Settings`, section `General`:
```

|         Field                    |   Value                                                   |
|----------------------------------|-----------------------------------------------------------|
| `Are you a customer or partner?` | `I'm an Okta customer adding an internal app`             |
| `App type`                       | `This is an internal app that we have created` - `CHECKED`|

Press Next to conclude the setup.

[Assign](https://help.okta.com/en-us/Content/Topics/Provisioning/lcm/lcm-assign-app-user.htm) the newly created app integration the user created in the previous step.

Under the `Sign On` tab of the newly created "Hosted Mender" locate the `View SAML setup instructions` button and click on it.

> Under the `Optional` there will be an XML containing the IdP metadata.

Reviewer 2: ```suggestion
Under the `Optional` section there will be an XML containing the IdP metadata.
```

> That contend needs to be copied to Hosted Mender.

Reviewer 2: ```suggestion
You need to copy that content to Hosted Mender.
```

## Setup on the Hosted Mender side

Copy the Okta IdP metadata generated in the previous step.

![Okta to mender](okta-idp-to-mender.png)

> Once the XML content has been copied, Mender will generate the required links.

Reviewer 2: ```suggestion
Once you copy the XML to it, Hosted Mender will generate the required links.
```

The `Entity ID` and `ACS URL` will be used to configure Okta.

> The `Start URL` is the URL to share with the user to log in once the setup is completed. 

Reviewer 2: ```suggestion
The `Start URL` is the URL to share with the user to log in once you complete the setup. 
```

![Okta to mender](mender-saml-url.png)

> Before going back to concluding the Okta setup with the correct values,
> in Hosted Mender, from the upper left dropdown select `User management` and click `CREATE NEW USER`.
> Give it the same email, assign it a role but **leave the password blank**.

Reviewer 2: ```suggestion
Before concluding the Okta setup with the correct values,
in Hosted Mender, from the upper left dropdown, select `User management` and click `Create new user`.
Then, give it the same email, and assign it a role, but **leave the password blank**.
```

![Mender create user](mender_create_user.png)

## Concluding the Okta setup

> Now its time to get back to Okta and replace the temporary urls with real ones.

Reviewer 2: ```suggestion
Now it's time to return to Okta and replace the temporary URLs with the real ones.
```

| Field (Okta)                  |  Field (Mender) |   Value          |
|-------------------------------|-----------------|------------------|
| `Audience URI (SP Entity ID)` | `Entity ID`     | `https://hosted.mender.io/api/management/v1/useradm/sso/sp/metadata/6651b207-f809-553d-a64a-821874acb9a7` |
| `Single sign-on URL`          | `ACS URL`       | `https://hosted.mender.io/api/management/v1/useradm/auth/sso/6651b207-f809-553d-a64a-821874acb9a7/acs` |

> Conclude the editing of the application config by not changing any other values.
> This concludes the setup on the Okta side.

Reviewer 2: ```suggestion
Don't change any other values and conclude the editing of the application.
This ends the setup on the Okta side.
```

## Accessing the login

> Copy the `Start URL` (third URL provided by Hosted Mender) into the browser an Okta login will pop up.
> Once complete the Okta authentication (you will have to install 'Okta Verify', the mobile app for Okta mfa) you will be redirected to Hosted Mender.

Reviewer 2: ```suggestion
Copy the `Start URL` (the third URL provided by Hosted Mender) into the browser, and an Okta login will pop up.
Once complete the Okta authentication (you will have to install 'Okta Verify', the mobile app for Okta MFA), you will be redirected to Hosted Mender.
```

> Once you complete with the Okta authentication (you will have to install 'Okta Verify', the mobile app for Okta MFA), you will be redirected to Hosted Mender.

Reviewer: Maybe a short additional section here how the login/authentication flow will work from now on in the standard case?
